src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorLSPTextViewConnectionListener.cs

```diff
@@ -259,16 +254,69 @@ private static bool CheckRazorServerCapability(JToken token)
             return isRazorLanguageServer;
         }
 
-        private static EditorSettings GetRazorEditorOptions(IVsTextManager2 textManager)
+        private static EditorSettings InitializeRazorTextViewOptions(IVsTextManager4 textManager, RazorEditorOptionsTracker optionsTracker)
+        {
+            var insertSpaces = RazorLSPOptions.Default.InsertSpaces;
+            var tabSize = RazorLSPOptions.Default.TabSize;
+
+            var langPrefs3 = new LANGPREFERENCES3[] { new LANGPREFERENCES3() { guidLang = RazorLSPConstants.RazorLanguageServiceGuid } }; ;
+            if (VSConstants.S_OK == textManager.GetUserPreferences4(null, langPrefs3, null))
+            {
+                // General options
+                optionsTracker.ViewOptions.SetOptionValue(DefaultTextViewOptions.UseVirtualSpaceName, Convert.ToBoolean(langPrefs3[0].fVirtualSpace));
+
+                var wordWrapStyle = WordWrapStyles.None;
+                if (Convert.ToBoolean(langPrefs3[0].fWordWrap))
+                {
+                    wordWrapStyle |= WordWrapStyles.WordWrap;
+                    if (Convert.ToBoolean(langPrefs3[0].fWordWrapGlyphs))
+                    {
+                        wordWrapStyle |= WordWrapStyles.VisibleGlyphs;
+                    }
+                }
+
+                optionsTracker.ViewOptions.SetOptionValue(DefaultTextViewOptions.WordWrapStyleName, wordWrapStyle);
```

Comment:
WOwz there's a lot here!! 😄  How'd you find all these out of curiosity?

Kinda nuts how arcane all this is for language services. Are we unique among other languages? Also, is there any sort of helpers that the editor offers to manage all this?

Reply:
Mostly trial-and-error - I noticed that `LANGPREFERENCES2` was missing some of the options from Tools->Options, so I was thinking if there's a `2`, there might be a `LANGUAGEPREFERENCES3`. 😁 Then I manually matched up the property names from the class with the Tools->Options option names, and at the end went tested each to confirm that my original match assumptions were correct.

We're unique among languages since we override the HTML editor. Most of these options are [set](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Editor/VisualStudio/Impl/ViewAdapter/SimpleTextViewWindow.cs&version=GBmain&line=1532&lineEnd=1532&lineStartColumn=1&lineEndColumn=47&lineStyle=plain&_a=contents) when the HTML editor is created. Since when this occurs we haven't yet overridden the editor with our language GUID, our options are also not correctly configured. I'm not aware of any helpers the editor offers to manage this unfortunately, I think we might be the only ones with a problem since we override the editor. :(

Reply:
Ah interesting, would it make sense for our [SetLanguageServiceId](https://github.com/dotnet/aspnetcore-tooling/blob/c8f58ccaf4241e8d14135593744b482514e4cc8e/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorLSPTextViewConnectionListener.cs#L118) call to own this in the future? Wondering if this also reveals a bug in the CoreEditor's handling of SetLanguageServiceId is all 😄.

Not saying to block this PR on CoreEditor but just throwing out some ideas for the future / follow up

Reply:
Good point, I think it makes more sense for it to go within the `SetLanguageServiceId` call. I filed an issue on Editor here:
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1408493